src/circuit/gadget/ecc/chip.rs

```diff
@@ -0,0 +1,362 @@
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+/// For each Orchard fixed base, we precompute:
+/// - coefficients for x-coordinate interpolation polynomials, and
+/// - z-values such that y + z = u^2 some square while -y + z is non-square.
+pub struct EccLoaded<C: CurveAffine> {
+    lagrange_coeffs: Option<HashMap<OrchardFixedBases<C>, Vec<Vec<C::Base>>>>,
+    lagrange_coeffs_short: Option<HashMap<OrchardFixedBases<C>, Vec<Vec<C::Base>>>>,
+    z: Option<HashMap<OrchardFixedBases<C>, [u64; 85]>>,
+    z_short: Option<HashMap<OrchardFixedBases<C>, [u64; 22]>>,
+}
+
+impl<C: CurveAffine> EccLoaded<C> {
+    fn get_lagrange_coeffs(&self, point: OrchardFixedBases<C>) -> Option<Vec<Vec<C::Base>>> {
+        if let Some(lagrange_coeffs) = &self.lagrange_coeffs {
+            lagrange_coeffs.get(&point).cloned()
+        } else {
+            None
+        }
+    }
+
+    fn get_lagrange_coeffs_short(&self, point: OrchardFixedBases<C>) -> Option<Vec<Vec<C::Base>>> {
+        if let Some(lagrange_coeffs_short) = &self.lagrange_coeffs_short {
+            lagrange_coeffs_short.get(&point).cloned()
+        } else {
+            None
+        }
+    }
+
+    fn get_z(&self, point: OrchardFixedBases<C>) -> Option<[u64; 85]> {
+        if let Some(z) = &self.z {
+            z.get(&point).cloned()
+        } else {
+            None
+        }
+    }
+
+    fn get_z_short(&self, point: OrchardFixedBases<C>) -> Option<[u64; 22]> {
+        if let Some(z_short) = &self.z_short {
+            z_short.get(&point).cloned()
+        } else {
+            None
+        }
+    }
+}
+
+impl<C: CurveAffine> FixedPoints<C> for OrchardFixedBases<C> {}
+
+impl<C: CurveAffine> Chip for EccChip<C> {
+    type Config = EccConfig;
+    type Field = C::Base;
+    type Loaded = EccLoaded<C>;
+
+    fn load(layouter: &mut impl Layouter<Self>) -> Result<Self::Loaded, Error> {
+        // Load fixed bases (interpolation polynomials)
+        todo!()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct CellValue<F: FieldExt> {
+    cell: Cell,
+    value: Option<F>,
+}
+
+impl<F: FieldExt> CellValue<F> {
+    pub fn new(cell: Cell, value: Option<F>) -> Self {
+        CellValue { cell, value }
+    }
+}
+
+#[derive(Clone, Debug)]
+/// A scalar used for variable-base scalar multiplication. This is decomposed
+/// bitwise in big-endian order, i.e. [k_n, k_{n-1}, ..., k_0] where
+/// `scalar = k_0 + k_1 * 2 + ... + k_n * 2^n`.
+pub struct EccScalarVar<C: CurveAffine> {
+    value: Option<C::Scalar>,
+    k_bits: Vec<CellValue<C::Base>>,
+}
+
+/// A full-width scalar used for variable-base scalar multiplication.
+/// This is decomposed in chunks of `window_width` bits in little-endian order.
+/// For example, if `window_width` = 3, we will have [k_0, k_1, ..., k_n]
+/// where `scalar = k_0 + k_1 * (2^3) + ... + k_n * (2^3)^n`.
+#[derive(Clone, Debug)]
+pub struct EccScalarFixed<C: CurveAffine> {
+    value: Option<C::Scalar>,
+    k_bits: Vec<CellValue<C::Base>>,
```

Comment:
(nit) Since the bits in the decomposition have to be consecutive, we could represent them with a `CellValueRange` data structure